src/esbuild.ts

```diff
@@ -0,0 +1,130 @@
+import path from "path";
+import { globSync } from "glob";
+import { BuildOptions, Loader, Plugin } from "esbuild";
+import { Args } from "./args.js";
+import esbuildPlugin, { PluginOptions, defaults as pluginDefaults } from "./esbuild-plugin.js";
+
+export interface EsbuildOptions extends Partial<BuildOptions> {
+  plugins: Plugin[];
+}
+
+const loader: { [ext: string]: Loader } = {
+  ".tsx": "tsx",
+  ".ts": "ts",
+  ".js": "js",
+  ".jsx": "jsx",
+  ".json": "json",
+  ".png": "file",
+  ".jpg": "file",
+  ".jpeg": "file",
+  ".gif": "file",
+  ".svg": "file",
+  ".woff": "file",
+  ".woff2": "file",
+  ".otf": "file",
+  ".eot": "file",
+  ".ttf": "file",
+};
+
+const entryPointExtensions = "app.{js,ts,mjs,mts,tsx,jsx}";
+// FIXME: make cross platform
+const entryPointsMatcher = /(app\/assets\/js\/|slices\/(.*\/)assets\/js\/)/;
+
+const findEntryPoints = (root: string): Record<string, string> => {
+  const result: Record<string, string> = {};
+
+  // TODO: should this be done explicitly within the root?
+  const entryPoints = globSync([
+    path.join("app", "assets", "js", "**", entryPointExtensions),
+    path.join("slices", "*", "assets", "js", "**", entryPointExtensions),
+  ]);
+
+  entryPoints.forEach((entryPoint) => {
+    let modifiedPath = entryPoint.replace(entryPointsMatcher, "$2");
+    const relativePath = path.relative(root, modifiedPath);
+
+    const { dir, name } = path.parse(relativePath);
+
+    if (dir) {
+      modifiedPath = path.join(dir, name);
+    } else {
+      modifiedPath = name;
+    }
+
+    result[modifiedPath] = entryPoint;
+  });
+
+  return result;
+};
+
+// TODO: feels like this really should be passed a root too, to become the cwd for globSync
```

Comment:
This was the other TODO I added here. The one other place in this file where we were doing file operations without providing an explicit root.

Reply:
@timriley IIRC, by adding the absolute path, Esbuild will report the absolute path in the metadata, so it's a bit cumbersome to work lately with them in the context of manifest generation.

Reply:
Ah, this is useful to know, thanks @jodosha!

The only downside of the current arrangement is that it's kind of contingent of the cwd also equalling the app root. In 99% of usage, I don't expect these things will diverge, but the fact that we're allowing a `root:` to be provided to `assets.run` means that some of these globs that reply on the cwd might start to fail.

However, I'm comfortable enough with us leaving this as a known gap for the moment. We can come back to solidify this aspect later.

And perhaps we don't document the `assets.run` `root:` option as "public" for now, too.